File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(!head) return false;
        ListNode *slow = head->next;
        ListNode *fast;
        if(slow) {
            fast = slow->next;
        } else {
            return false;
        }
        while(fast && slow) {
            if(fast == slow) return true;
            slow = slow->next;
            if(fast->next) {
                fast = fast->next->next;
            } else {
                return false;
            }
        }
        return false;
    }
};

File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head:
            return False
        fast = head.next
        slow = head

        while fast and slow:
            if fast == slow:
                return True
            if not fast.next:
                break
            fast = fast.next.next
            slow = slow.next
        return False
